agent/src/discovery_handler_manager/discovery_handler_registry.rs

```diff
@@ -169,35 +171,60 @@ pub trait DiscoveryHandlerRegistry: Sync + Send {
 /// Real world implementation of the Discovery Handler Request
 struct DHRequestImpl {
     endpoints: RwLock<Vec<watch::Receiver<Vec<Arc<DiscoveredDevice>>>>>,
-    notifier: watch::Sender<Vec<Arc<DiscoveredDevice>>>,
+    notifier: watch::Sender<crate::device_manager::cdi::Kind>,
     key: String,
     handler_name: String,
     details: String,
     properties: Vec<DiscoveryProperty>,
-    extra_device_properties: HashMap<String, String>,
+    extra_device_properties: RwLock<HashMap<String, String>>,
     kube_client: Arc<dyn DiscoveryManagerKubeInterface>,
     termination_notifier: Arc<Notify>,
 }
 
+#[async_trait]
 impl DiscoveryHandlerRequest for DHRequestImpl {
-    fn get_instances(&self) -> Result<Vec<Instance>, DiscoveryError> {
+    async fn get_instances(&self) -> Result<Vec<Instance>, DiscoveryError> {
+        let properties = self.extra_device_properties.read().await;
```

Comment:
Same point here: could this skipped and instead let the called function (`device_to_instance`) access the properties? Or is the concern wanting to avoid too many reads? It is RWLock, so reads should be cheaper. I could also see how your strategy is trying to avoid a racecase where some containers have properties that were written after the initial call

Reply:
acquiring the lock is an async operation, you cannot do this from within the `map` call (same reason for not doing it in `device_to_instance`). 
But in the end I find it better anyway, as it means we have a consistent set of properties for the entire run

Reply:
`device_to_instance` could be made async but i am aligned on wanting the consistent set of properties